lib/processors/manifestEnhancer.js

```diff
@@ -0,0 +1,417 @@
+				// in case it didn't existing before.
+				if (!modelConfig.settings) {
+					modelConfig.settings = settings;
+				}
+			})
+		);
+	}
+
+	async processSapUi5LibraryI18n() {
+		let sapUi5LibraryI18n = this.manifest["sap.ui5"]?.library?.i18n;
+
+		// Process enhanceWith bundles first, as they check for an existing supportedLocales property
+		// defined by the developer, but not the one generated by the tooling.
+		await this.processTerminologiesAndEnhanceWith(sapUi5LibraryI18n);
+
+		const i18nBundleUrl = getBundleUrlFromSapUi5LibraryI18n(sapUi5LibraryI18n);
+		if (i18nBundleUrl && !sapUi5LibraryI18n?.supportedLocales) {
+			const supportedLocales = await this.getSupportedLocales(i18nBundleUrl, sapUi5LibraryI18n?.fallbackLocale);
+			if (supportedLocales.length > 0) {
+				if (!sapUi5LibraryI18n || typeof sapUi5LibraryI18n !== "object") {
+					this.manifest["sap.ui5"] ??= {};
+					this.manifest["sap.ui5"].library ??= {};
+					sapUi5LibraryI18n = this.manifest["sap.ui5"].library.i18n = {
+						bundleUrl: i18nBundleUrl
+					};
+				}
+				sapUi5LibraryI18n.supportedLocales = supportedLocales;
+				this.markModified();
+			}
+		}
+	}
+
+	async run() {
+		// Prevent multiple invocations
+		if (this.runInvoked) {
+			throw new Error("ManifestEnhancer#run can only be invoked once per instance");
+		}
+		this.runInvoked = true;
+
+		if (!this.manifest._version) {
+			log.verbose("manifest.json: _version is not defined. No supportedLocales are generated");
+			return;
+		}
+
+		if (lt(this.manifest._version, APP_DESCRIPTOR_V22)) {
+			log.verbose("manifest.json: _version is lower than 1.21.0 so no supportedLocales can be generated");
+			return;
+		}
+
+		if (this.manifest["sap.app"].type === "library") {
+			await this.processSapUi5LibraryI18n();
+		} else {
+			await Promise.all([
+				this.processSapAppI18n(),
+				this.processSapUi5Models()
+			]);
+		}
+
+		if (this.isModified) {
+			return this.manifest;
+		}
+	}
+}
+
+/**
+ * @module @ui5/builder/processors/manifestEnhancer
+ */
+
+/**
+ * Enriches the content of the manifest.json file.
+ *
+ * @public
+ * @function default
+ * @static
+ *
+ * @param {object} parameters Parameters
+ * @param {@ui5/fs/Resource[]} parameters.resources List of manifest.json resources to be processed
+ * @param {fs|module:@ui5/fs/fsInterface} parameters.fs Node fs or custom
+ *    [fs interface]{@link module:@ui5/fs/fsInterface}.
+ * @returns {Promise<Array<@ui5/fs/Resource|undefined>>} Promise resolving with an array of modified resources
+ */
+export default async function({resources, fs}) {
+	return Promise.all(
+		resources.map(async (resource) => {
+			const manifest = await resource.getString();
+			const filePath = resource.getPath();
+			const manifestEnhancer = new ManifestEnhancer(manifest, filePath, fs);
+			const enrichedManifest = await manifestEnhancer.run();
+			if (enrichedManifest) {
+				resource.setString(JSON.stringify(enrichedManifest, null, 2));
+			}
+			return resource;
+		})
+	);
```

Comment:
I think we should only return a resource if it has been modified. Maybe like this?

```suggestion
	const res = await Promise.all(
		resources.map(async (resource) => {
			const manifest = await resource.getString();
			const filePath = resource.getPath();
			const manifestEnhancer = new ManifestEnhancer(manifest, filePath, fs);
			const enrichedManifest = await manifestEnhancer.run();
			if (enrichedManifest) {
				resource.setString(JSON.stringify(enrichedManifest, null, 2));
				return resource;
			}
		})
	);
	return res.filter(($) => $);
```